content/blog/2022-04-07-compiling-black-with-mypyc-part-1.md

```diff
@@ -0,0 +1,591 @@
+creating build/lib.linux-x86_64-3.8
+gcc -pthread -shared build/temp.linux-x86_64-3.8/build/__native.o -o build/lib.linux-x86_64-3.8/coefficient_finder.cpython-38-x86_64-linux-gnu.so
+
+$ python -c "import coefficient_finder"
+compute() took 331.6 milliseconds
+```
+
+And would you look at that, mypyc gave us a near 6x improvement over CPython! What's neat
+is that I didn't even need to type all of the variables thanks to mypy's type inference.
+
+> And actually, since the `mypyc` script is a simple wrapper that generates a setuptools
+> project and builds it in-place, you can peek into the `build` directory and look at the
+> generated C code (should be `__native.c`).
+
+Feel free to edit the program and play around with mypyc, seeing what little scripts it
+can speed up. Pro-tip, to restore back to the interpreted version of your program, you can
+delete the `.so` file on Linux & MacOS or the `.pyd` file on Windows. In this example
+I got `coefficient_finder.cpython-38-x86_64-linux-gnu.so`.
+
+______________________________________________________________________
+
+### It's not a silver bullet
+
+There's several downsides to using mypyc though. While the project aims for excellent
+compatibility with standard Python behaviour, there's some major deviations worth noting:
+
+- Types are enforced at runtime and violations **will** cause a TypeError (this makes
+  using `typing.Any` quite dangerous).
+
+- Compiled modules can't be run directly, you have to import 'em instead.
+
+- Monkeypatching anything compiled is unlikely to work as compiled code skips many of the
+  runtime `__dict__` lookups normal Python does.
+
+- Assignments to class and instance namespaces will either error or do nothing, in
+  particular you can't add previously undeclared attributes later on.
+
+- Standard multiple inheritance isn't supported.
+
+- Profilers (e.g. CProfile), debuggers (e.g. pdb), and tracing hooks (e.g. coverage) won't
+  work as the C code doesn't trigger the relevant hooks.
+
+There's more details available in the official
+[mypyc documentation here][mypyc-deviations-one] and [also here][mypyc-deviations-two].
+
+## black + mypyc: Initial steps
+
+I wasn't the first person to integrate mypyc into Black, way back in September 2019,
+[@msullivan] [opened a PR getting Black ready][original-pr] for compilation. Unfortunately
+as typical in open source projects, no one took the half-completed work and pushed it to
+production, i.e. PyPI, ... for almost two years.
+
+> You might be wondering why performance even matters, well clearly it mattered a lot
+> since [GH-366] was opened in June 2018! The TL;DR is that for environments where Black
+> is **run on save automatically, the more responsive it is the better** as less time
+> is spent in a laggy editor window.
+>
+> Start up time is important too given imports are costly, but this issue was explicitly
+> about formatting throughput (it turns out mypyc can reduce import time too!).
```

Comment:
Perhaps the parenthesized clause should be moved before the comma so that the information about imports reads as one thing.
> Start up time is important too given imports are costly (it turns out mypyc can reduce import time too!), but this issue was explicitly about formatting throughput.